### ISSUE TITLE
No hook for invalid challenge

### ISSUE THREAD

**Issue by RasmusLindroth** [josteink/le-godaddy-dns#4]
There is no hook for invalid challenge.

```
 + Godaddy hook executing: invalid_challenge
Traceback (most recent call last):
  File "/home/rasmus/letsencrypt/le-godaddy-dns/godaddy.py", line 93, in <module>
    main(sys.argv[1:])
  File "/home/rasmus/letsencrypt/le-godaddy-dns/godaddy.py", line 89, in main
    ops[argv[0]](argv[1:])
KeyError: 'invalid_challenge'
```
See this issue for a solution.
https://github.com/kappataumu/letsencrypt-cloudflare-hook/issues/26

Thank you for a great script :)

**Comment by josteink** [comment]
Attempted fixed in: https://github.com/josteink/le-godaddy-dns/commit/454ee15fae4b6eb6ce9982f010edc32ff6ae5707

Please let me know if this solves your issues :)

**Comment by RasmusLindroth** [comment]
Thank you for the fast response :) I think that should solve that part, but now I got an other error.

```
 + Godaddy hook executing: request_failure
Traceback (most recent call last):
  File "/home/rasmus/letsencrypt/le-godaddy-dns/godaddy.py", line 102, in <module>
    main(sys.argv[1:])
  File "/home/rasmus/letsencrypt/le-godaddy-dns/godaddy.py", line 98, in main
    ops[argv[0]](argv[1:])
KeyError: 'request_failure'
```
I will be out of town for this weekend so I can take a closer look after that. Have a good weekend!

**Comment by josteink** [comment]
Any news on this issue?

**Comment by RasmusLindroth** [comment]
The hook for invalid challange work, now its just the validation that only work sometimes. I have to run the script multiple times to get all challenges validated. I can't pinpoint the problem, but it seems like the records don't have been updated sometimes like here.

```
 + Godaddy hook executing: deploy_challenge
 + Updating TXT record for _acme-challenge.rasmus.xyz to 'CkKpLjBnvuaoYfnFAagJ5xIAeAMWCzXIpDZazWey0HQ'.
 + Responding to challenge for rasmus.xyz...
 + Godaddy hook executing: clean_challenge
 + Updating TXT record for _acme-challenge.rasmus.xyz to 'null'.
 + Godaddy hook executing: invalid_challenge
 + invalid challenge for domain rasmus.xyz: {
  "type": "dns-01",
  "status": "invalid",
  "error": {
    "type": "urn:acme:error:unauthorized",
    "detail": "No TXT records found for DNS challenge",
    "status": 403
  },
```
But that maybe is a problem with godaddy? 

But this project still needs a hook for `request_failure` or else the script gets a key error, the fix should be similar to 454ee15.

Thank you for your help!

**Comment by josteink** [comment]
> But that maybe is a problem with godaddy?

I've applied the following patch to the instance I'm running myself.

````diff
diff --git a/godaddy.py b/godaddy.py
index 61b7ba2..844b966 100755
--- a/godaddy.py
+++ b/godaddy.py
@@ -58,10 +58,13 @@ def create_txt_record(args):
 
 
 def delete_txt_record(args):
+    from time import sleep
     domain = args[0]
     # using client.delete_record() is dangerous. null it instead!
     # https://github.com/eXamadeus/godaddypy/issues/13
     _update_dns(domain, "null")
+    logger.info(" + Domain cleared. Sleeping to ensure it's effective!")
+    sleep(10)
````

Does that help for you? If so, maybe we should make it mainline code.

> But this project still needs a hook for request_failure or else the script gets a key error, the fix should be similar to 454ee15.

Feel free to issue a PR for that, and I'll gladly merge it.

**Comment by RasmusLindroth** [comment]
I have made a PR for request_failure, so now the script exists nicely even if there is an error. Unfortunate the sleep patch isn't working for me, but it isn't a real problem because I only need to run the validation 1-2 more times and it works. I will come back if I figure something out, but this issue is solved :)

PR: https://github.com/josteink/le-godaddy-dns/pull/6

**Comment by josteink** [comment]
Closed via #6.
